R/tbl_listing.R

```diff
@@ -0,0 +1,126 @@
+#' @param keys ([`tidy-select`][dplyr::dplyr_tidy_select])\cr
+#'   columns to be highlighted on the left of the listing.
+#' @param order_by ([`tidy-select`][dplyr::dplyr_tidy_select])\cr
+#'   columns to be sorted. It defaults to `keys`.
+#' @param do_not_print_duplicated_keys (`logical`)\cr
+#'   whether to add blank values where key columns have duplicate values. Defaults to `TRUE`.
+#'
+#' @details
+#' This function converts a data frame into a listing format and it relies on the user's pre-processing
+#' to decide which column to show and select unique or non-unique rows. Also sorting depends on the user
+#' pre-processing.
+#'
+#' @note
+#' `NA` values are not printed by default in `{gtsummary}`. Please consider making them explicit if
+#' you want them to be displayed in the listing.
+#'
+#' @examples
+#' # Example 1 --------------------------------
+#' trial_data <- trial |>
+#'   dplyr::select(trt, age, marker, stage)
+#'
+#' tbl_listing(trial_data, keys = c(trt, stage))
+#'
+#' @export
+tbl_listing <- function(data,
+                        keys = NULL,
+                        order_by = keys,
+                        do_not_print_duplicated_keys = TRUE) {
+  set_cli_abort_call()
+
+  # Checks -----------------------------------
+  check_not_missing(data)
+  check_data_frame(data)
+  check_scalar_logical(do_not_print_duplicated_keys)
+
+  # Process arguments ------------------------
+  cards::process_selectors(data, keys = {{ keys }})
+  cards::process_selectors(data, order_by = {{ order_by }})
+
+  tbl_listing_inputs <- as.list(environment())
+
+  # Sorting ----------------------------------
+  if (!is.null(order_by)) {
+    data <- data |>
+      dplyr::arrange(across(all_of(order_by)))
+
+      # Inform about happened sorting if interactive
+      if (interactive()) {
+        main_message <-
+          if (identical(order_by, keys)) {
+            "Sorting incoming data by key columns."
+          } else {
+            "Sorting incoming data by column{?s} {.val {order_by}}."
+          }
+
+        cli::cli_inform(
+          c(
+            "v" = main_message,
+            "i" = "If you want to change the sorting, please sort the data frame before passing it to `tbl_listing()`."
+          ),
+        call = get_cli_abort_call()
+        )
+      }
+  }
+
+  # Reorder the full set of cols to ensure key columns are first
+  ordercols <- c(keys, setdiff(names(data), keys))
+  data <- data[, ordercols]
+
+  # Build the listing ----------------------------------------------------------
+  x <- gtsummary::as_gtsummary(data)
+  x$inputs <- tbl_listing_inputs
+
+  # return with class and attributes -------------------------------------------
+  structure(
+    x,
+    class = c("tbl_listing", "gtsummary")
+  )
+}
+
+
+#' @param x (`tbl_listing`)\cr
+#'   a `tbl_listing` object to be printed.
+#' @inheritParams gtsummary::print.gtsummary
+#'
+#' @name tbl_listing
+#' @exportS3Method
+print.tbl_listing <- function(x,
```

Comment:
@ddsjoberg this allows the following:
```r
devtools::load_all(".")
trial_df <- trial |>
  dplyr::select(trt, age, marker, stage)

a_lst <- tbl_listing(trial_df, keys = c(trt, stage))
a_lst$table_body # Still has the values

b_lst <- tbl_split_by_rows(a_lst, row_numbers = 5)
b_lst[[2]] # printing it presents the first value for a random line split!!!
```

Reply:
I really like this solution!!

Reply:
I think it would work also for docx output right? because it is done before as_flex_table

Reply:
add_blank_rows works on this too though (not on the variable of course but row_number yes)